### PR TITLE
[WIP] Doc - Intercept computation

### PR DESCRIPTION
**Pull request by Klopfe**
PR for issue #70, on writing doc for the computation of the intercept in skglm.

Started a tex document doc/intercept.tex

### CODE REVIEW

File: doc/intercept.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+$$
+\begin{align}
+    \nabla_{\beta_0}F(X\beta^\star + \beta_0^\star\boldsymbol{1}_{n}) = 0
+    \enspace .
+\end{align}
+$$
+Moreover, we have that 
+$$
+\begin{align}
+    \nabla_{\beta_0}F(X\beta^\star + \beta_0^\star\boldsymbol{1}_{n}) = \boldsymbol{1}_{n}^\top \nabla F(X\beta^\star + \beta_0^\star\boldsymbol{1}_{n})
```
Comment (PABannier) [inline doc/intercept.md:44]: Just to make sure, maybe add a subscript beta for nabla on the rhs of the equation. The grad on rhs is wrt to beta, right?

File: doc/intercept.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+We will now derive the update used in Equation 2 for three different datafitting functions. 
+
+---
+
+## The Quadratic datafit
+
+We define 
+$$
+\begin{align}
+    F(X\beta + \beta_0\boldsymbol{1}_{n})) = \frac{1}{2n} \lVert y - X\beta + \beta_0\boldsymbol{1}_{n} \rVert^2_2
```
Comment (PABannier) [inline doc/intercept.md:58]: double parenthesis on lhs

Comment (PABannier) [inline doc/intercept.md:58]: on rhs, isn't it -beta_0, am I missing something?

File: doc/intercept.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+\end{align}
+$$
+In this case $\nabla F(z) = \frac{1}{n}(z - y)$ hence Eq. 4 is equal to:
+$$
+\begin{align}
+    \nabla_{\beta_0}F(X\beta + \beta_0\boldsymbol{1}_{n}) = \frac{1}{n}\sum_{i=1}^n(X_{i:}\beta + \beta_0 - y_i)
+    \enspace .
+\end{align}
+$$
+Finally, the Lispchitz constant $L_0 = \frac{1}{n}\sum_{i=1}^n 1^2 = 1$.
```
Comment (PABannier) [inline doc/intercept.md:69]: The Lipschitz constant for the intercept would upper-bound the spectral radius of the Hessian of the datafit w.r.t. to the intercept, if I understand correctly. Maybe make it a bit more explicit. (We take the derivative of grad_{beta_0} w.r.t. to beta_0 or something like this)..

File: doc/intercept.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+The quantity $\frac{1}{L_0}\nabla_{\beta_0}F(X\beta^{(k)} + \beta_0^{(k)}\boldsymbol{1}_{n})$ is called $\texttt{intercept\_update\_step}$ in the class $\texttt{Quadratic}$ of $\texttt{skglm}$.
+
+---
+
+## The Logistic datafit
+
+In this case, 
+$$
+\begin{align}
+    F(X\beta + \beta_0\boldsymbol{1}_{n}) = \frac{1}{n} \sum_{i=1}^n \log(1 + \exp(-y_i X_{i:}\beta))
```
Comment (PABannier) [inline doc/intercept.md:80]: i'd make beta_0 appear explicitly

File: doc/intercept.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+$$
+
+We can then write
+$$
+\begin{align}
+ \nabla_{\beta_0}F(X\beta + \beta_0\boldsymbol{1}_{n}) = \frac{1}{n} \sum_{i=1}^n  \frac{-y_i}{1 + \exp(- y_iX_{i:}\beta)} \enspace .
+\end{align}
+$$
+
+Finally, the Lispchitz constant $L_0 = \frac{1}{4n}\sum_{i=1}^n 1^2 = \frac{1}{4}$.
```
Comment (PABannier) [inline doc/intercept.md:91]: maybe add in a sentence that \frac{1}{4} comes from upper-bounding the hessian.

**Review by PABannier — COMMENTED** [review]
A few minor comments, but much clearer for me on my side. Thanks @Klopfe !

File: doc/intercept.md (hunk 2 above)
Comment (Klopfe) [reply doc/intercept.md:58]: no you're right 👍

File: doc/intercept.md (hunk 3 above)
Comment (Klopfe) [reply doc/intercept.md:69]: well here for coordinate descent it's that $\forall h\in \mathbb{R}, |\nabla f(x + he_j) - \nabla f(x)| \leq L_j |h|$ so it's not anymore a global bound on the spectral radius of the Hessian, it's a bound in one direction.

File: doc/intercept.md (hunk 1 above)
Comment (Klopfe) [reply doc/intercept.md]: right!

File: doc/intercept.md (hunk 3 above)
Comment (Klopfe) [reply doc/intercept.md]: I have added the inequality it satisfies. Let me know if you think it's enough. Deriving the formulas is quite technical.

**Review by mathurinm — APPROVED** [review]
Can you make this file show up on the doc ? Then go for merge

File: doc/intercept.md (hunk 3 above)
Comment (PABannier) [reply doc/intercept.md]: Ok LGTM then!

**Review by PABannier — APPROVED** [review]

**Comment by PABannier** [comment]
Thanks @Klopfe !

File: doc/conf.py [hunk 6]
```diff
@@ -58,9 +58,15 @@
     'sphinx.ext.autosectionlabel',
     'numpydoc',
     'sphinx.ext.linkcode',
-    'gh_substitutions',  # custom ext, see ./sphinxext/gh_substitutions.py
+    'gh_substitutions',
+    'myst_parser',
+    # custom ext, see ./sphinxext/gh_substitutions.py
```
Comment (mathurinm) [inline doc/conf.py:63]: can you keep this on the gh_substitutions line (it's a bit mysterious so it helps to know it's a custom one)

File: doc/doc-requirements.txt [hunk 7]
```diff
@@ -1,6 +1,7 @@
 benchopt
 libsvmdata>=0.2
 matplotlib>=2.0.0
+myst_parser
```
Comment (mathurinm) [inline doc/doc-requirements.txt:4]: nice find

**Review by mathurinm — APPROVED** [review]
just 1 small fix then go for merge when green

**Pull request merged**